File: backend/app/schemas/prediction.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class PredictionItem(BaseModel):
    horse_number: int
    horse_id: Optional[str] = None
    horse_name: Optional[str] = None
    predicted_rank: int
    probability: float
    odds: Optional[float] = None
    popularity: Optional[int] = None


class RecommendedBet(BaseModel):
    bet_type: str
    detail: str
    confidence: str


class PredictionResult(BaseModel):
    predictions: list[PredictionItem]
    recommended_bets: list[RecommendedBet]


class PredictionResponse(BaseModel):
    prediction_id: int
    race_id: str
    model_version: str
    created_at: datetime
    results: PredictionResult

    class Config:
        from_attributes = True


class CreatePredictionRequest(BaseModel):
    race_id: str


class HistoryBase(BaseModel):
    bet_type: str = Field(..., description="Bet type (単勝, 馬連, etc.)")
    bet_detail: str = Field(..., description="Bet detail (e.g., '5' or '1-3')")
    bet_amount: Optional[int] = Field(None, description="Bet amount in yen")


class CreateHistoryRequest(HistoryBase):
    prediction_id: int


class HistoryItem(HistoryBase):
    id: int
    prediction_id: int
    is_hit: Optional[bool] = None
    payout: Optional[int] = None
    created_at: datetime

    class Config:
        from_attributes = True


class HistoryRace(BaseModel):
    race_id: str
    date: str
    race_name: Optional[str] = None


class HistoryWithRace(HistoryItem):
    race: Optional[HistoryRace] = None


class HistorySummary(BaseModel):
    total_bets: int
    total_hits: int
    hit_rate: float
    total_bet_amount: int
    total_payout: int
    roi: float


class HistoryListResponse(BaseModel):
    total: int
    summary: HistorySummary
    history: list[HistoryItem]


class UpdateHistoryResultRequest(BaseModel):
    is_hit: bool
    payout: Optional[int] = None


class HistoryUpdateResponse(BaseModel):
    id: int
    is_hit: bool
    payout: Optional[int] = None
    message: str


class AccuracyStats(BaseModel):
    top1: float
    top3: float
    top5: float = 0.0


class AccuracyByCategory(BaseModel):
    top1: float
    top3: float


class StatsResponse(BaseModel):
    period: str
    accuracy: AccuracyStats
    by_grade: dict[str, AccuracyByCategory] = {}
    by_track: dict[str, AccuracyByCategory] = {}

File: frontend/src/types/prediction.ts
export interface PredictionHorse {
  horse_number: number;
  horse_id: string;
  horse_name?: string;
  predicted_rank: number;
  probability: number;
  score?: number;
  odds?: number;
  popularity?: number;
  tansho_ev?: number;  // 単勝期待値
}

export interface RecommendedBet {
  bet_type: string;
  detail: string;
  confidence: 'high' | 'medium' | 'low';
  horse_name?: string;
  horse_names?: string[];  // 馬連用
  expected_value?: number;  // 期待値
  probability?: number;
  odds?: number;
}

export interface PredictionResult {
  predictions: PredictionHorse[];
  recommended_bets: RecommendedBet[];
  model_type: 'ml' | 'baseline';
}

export interface Prediction {
  prediction_id: number;
  race_id: string;
  model_version: string;
  created_at: string;
  results: PredictionResult;
}

export interface History {
  id: number;
  prediction_id: number;
  bet_type: string;
  bet_detail: string;
  bet_amount?: number;
  is_hit?: boolean;
  payout?: number;
  created_at: string;
}

export interface HistorySummary {
  total_bets: number;
  total_hits: number;
  hit_rate: number;
  total_bet_amount: number;
  total_payout: number;
  roi: number;
}

export interface HistoryListResponse {
  total: number;
  items: History[];
  summary: HistorySummary;
}

export interface AccuracyStats {
  period: string;
  total_predictions: number;
  win_accuracy: number;
  show_accuracy: number;
  avg_top3_hit: number;
}

export interface ScrapeStats {
  total_races: number;
  total_entries: number;
  total_horses: number;
  total_jockeys: number;
  total_predictions: number;
  latest_race_date?: string;
}
